primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Trait definitions for a Merkle tree gadget.
+
+use crate::merkle_tree::MerkleTreeScheme;
+use ark_ff::PrimeField;
+use jf_relation::{errors::CircuitError, BoolVar, Variable};
+
+mod rescue_merkle_tree;
+
+/// Circuit variable for a leaf element.
+#[derive(Debug, Clone)]
+pub struct LeafVar {
+    /// Position of the leaf element in the MT. Serves as UID.
+    pub uid: Variable,
+    /// The value of the leaf element.
+    pub elem: Variable,
+}
+
```

Comment:
Do you think we should make this an associated type of `trait MerkleTreeGadget` instead, just like `MerklePathVar`?
The main rationale being that some MT's leaf (e.g. SMT) doesn't have to care about or even have `uid`. So it may suggest leaving the concrete instantiation to specific gadgets.

Reply:
on this note, I noticed our current [`RescueSMT`'s `digest_leaf`](https://github.com/EspressoSystems/jellyfish/blob/main/primitives/src/merkle_tree/prelude.rs#L55) is actually hashing `[0, pos, elem]`, which I guess doesn't hurt, but when we treat it as a set (as we do when using SMT), we don't need that `pos`, no? @mrain

Reply:
No. In case of set, `elem` is of type `()`. Information is stored in `pos` actually.

Reply:
oh, right!

my original comment on making it an associated type might still be worth considering.

Reply:
Yes, good suggestion. Done in https://github.com/EspressoSystems/jellyfish/pull/158/commits/7288912d5ed7b6bd92b4e4e4ac9a8537cdf12b3c.